src/Orleans.Runtime.Abstractions/Streams/SiloPersistentStreamConfigurator.cs

```diff
@@ -49,6 +50,7 @@ public SiloPersistentStreamConfigurator(string name, Action<Action<IServiceColle
             ConfigureComponent<ILifecycleParticipant<ISiloLifecycle>>(PersistentStreamProvider.ParticipateIn<ISiloLifecycle>);
             ConfigureComponent<IQueueAdapterFactory>(adapterFactory);
             ConfigureComponent<IConfigurationValidator>(PersistentStreamStorageConfigurationValidator.Create);
+            base.configureDelegate(services => services.TryAddSingleton<IStreamFailureHandler>(NoOpStreamDeliveryFailureHandler.Create));
```

Comment:
should this be Singleton or transient? Before this pr,  the behavior is more like a transient, meaning stream provider cache the handler factory, and create a new handler each time the pulling agent asks for one.  And also , before this pr, failure handler can be initialized with an asynchronously, see `Func<string, Task<IStreamFailureHandler>> StreamFailureHandlerFactory`  

I bring this up because this is more than a compile time breaking change, it is also a functionality/behavior  breaking change,

Reply:
For the specific impl "NoOpStreamDeliveryFailureHandler", a singleton is safe.

In the case of a user configured handler, all configured components have been singletons so far, and I think that's fine for this one as well.  I understand that this is a deviation from the previous pattern, but I suspect most implementations will not contain per queue state.  If they do need per queue state, the application developer can register their impl as a transient named service.

Regarding the loss of asynchronous construction, if an impl needs asynchronous initialization, the pattern we've used in other components (like grain storage) is for them to subscribe to the silo lifecycle and initialize themselves at startup, rather than waiting until they're resolved from the container.  I don't think maintaining the previous asynchronous construction (for compatibility reasons) is sufficient to justify deviating from the common pattern we've settled on in other areas of the system for the following reasons.
- I don't think most users have even implemented a failure handler due to the difficulty of doing so prior to this change, so this will be a breaking change for very few.
- I don't think asynchronous initialization will be necessary very often.
- If asynchronous initialization is necessary, it's better to perform it at silo startup than when the component is resolved from the container, as we have better testing and recovery patterns around silo initialization problems.